mem-store.ts

```diff
@@ -91,124 +94,74 @@ function mem_store(this: any, options: any) {
+          seneca.fail('entity-id-exists', { type: ent.entity$, id: mement.id })
+          return
         }
 
         mement = seneca.util.deep(mement)
-        if (shouldMerge) {
+        const should_merge = Intern.should_merge(ent, options)
+
+        if (should_merge) {
           mement = Object.assign(prev || {}, mement)
         }
 
-        return upsertIfRequested(msg, function(err: any, ctx: UpsertIfRequestedContext): any {
-          if (err) {
-            return reply(err)
-          }
-
-          const { did_update } = (ctx as any)
-
-          if (did_update) {
-            const { out } = (ctx as any)
-            return reply(null, out)
-          }
-
-
-          prev = entmap[base][name][mement.id] = mement
-
-          seneca.log.debug(function() {
-            return [
-              'save/' + (isNewEntityBeingCreated(msg) ? 'insert' : 'update'),
-              ent.canon$({ string: 1 }),
-              mement,
-              desc,
-            ]
-          })
-
-          return reply(null, ent.make$(prev))
-        })
-
-
-        type UpsertIfRequestedContext = { did_update: boolean, out?: any } | undefined
-
-        type UpsertIfRequestedCallback = (
-          err: Error | null,
-          result?: UpsertIfRequestedContext
-        ) => any
-
-        function upsertIfRequested(msg: any, reply: UpsertIfRequestedCallback) {
-          // This is the query passed to the .save$ method.
-          //
-          const query_for_save = msg.q
-
-
-          if (isNewEntityBeingCreated(msg) && Array.isArray(query_for_save.upsert$)) {
-            const upsert_on = Common.clean(query_for_save.upsert$)
-
-
-            if (upsert_on.length > 0) {
-              if (!(base in entmap)) {
-                return reply(null, { did_update: false, out: null })
-              }
-
-              if (!(name in entmap[base])) {
-                return reply(null, { did_update: false, out: null })
-              }
-
-
-              const collection = entmap[base][name]
-              const docs = Object.values(collection)
-              const public_entdata = msg.ent.data$(false)
-
-
-              const doc_to_update = docs.find((doc: any) => {
-                return upsert_on.every((p: string) => {
-                  return p in public_entdata && public_entdata[p] === doc[p]
+        if (Intern.is_upsert_requested(msg)) {
+          const upsert_on = Intern.clean_array(msg.q.upsert$)
+
+          if (0 < upsert_on.length) {
+            const public_entdata = ent.data$(false)
+            const may_match = upsert_on.every((p: string) => p in public_entdata)
+
+            if (may_match) {
+              const match_by = upsert_on.reduce((h: any, p: string) => {
+                h[p] = public_entdata[p]
+                return h
+              }, {})
+
+              const updated_ent = Intern.update_ent(entmap, ent, match_by, public_entdata)
+
+              if (updated_ent) {
+                // TODO: DRY up.
```

Comment:
yes, definitely - this needs to be fixed
how will we solve this? how will we structure the callback functions?